SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -0,0 +1,235 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Management;
+using System.Net.NetworkInformation;
+using System.Text;
+using Microsoft.Win32;
+
+namespace SteamKit2
+{
+    abstract class MachineInfoProvider
+    {
+        public static MachineInfoProvider GetProvider()
+        {
+            switch ( Environment.OSVersion.Platform )
+            {
+                case PlatformID.Win32NT:
+                case PlatformID.Win32Windows:
+                    return new WindowsInfoProvider();
+
+                case PlatformID.Unix:
+                    if ( !Utils.IsRunningOnDarwin() )
+                    {
+                        return new LinuxInfoProvider();
+                    }
+                    break;
+            }
+
+            return new DefaultInfoProvider();
+        }
+
+        public abstract byte[] GetMachineGuid();
+        public abstract byte[] GetMacAddress();
+        public abstract byte[] GetDiskId();
+    }
+
+    class DefaultInfoProvider : MachineInfoProvider
+    {
+        public override byte[] GetMachineGuid()
+        {
+            return Encoding.UTF8.GetBytes( Environment.MachineName + "-SteamKit" );
+        }
+
+        public override byte[] GetMacAddress()
+        {
+            // mono seems to have a pretty solid implementation of NetworkInterface for our platforms
+            // if it turns out to be buggy we can always roll our own and poke into /sys/class/net on nix
+
+            var firstEth = NetworkInterface.GetAllNetworkInterfaces()
+                .Where( i => i.NetworkInterfaceType == NetworkInterfaceType.Ethernet )
+                .FirstOrDefault();
+
+            if ( firstEth == null )
+            {
+                // well...
+                return Encoding.UTF8.GetBytes( "SteamKit-MacAddress" );
+            }
+
+            return firstEth.GetPhysicalAddress().GetAddressBytes();
+        }
+
+        public override byte[] GetDiskId()
+        {
+            return Encoding.UTF8.GetBytes( "SteamKit-DiskId" );
+        }
+    }
+
+    class WindowsInfoProvider : DefaultInfoProvider
+    {
+        public override byte[] GetMachineGuid()
+        {
+            RegistryKey localKey = RegistryKey
+                .OpenBaseKey( Microsoft.Win32.RegistryHive.LocalMachine, RegistryView.Registry64 )
```

Comment:
https://msdn.microsoft.com/en-us/library/microsoft.win32.registryview%28v=vs.110%29.aspx

"If you request a 64-bit view on a 32-bit operating system, the returned keys will be in the 32-bit view."